python/ray/data/impl/lazy_block_list.py

```diff
@@ -25,122 +64,380 @@ class LazyBlockList(BlockList):
+        assert len(tasks) == len(self._block_partitions), (
+            tasks,
             self._block_partitions,
         )
+        assert len(tasks) == len(self._block_partitions_meta), (
+            tasks,
+            self._block_partitions_meta,
+        )
+        assert len(tasks) == len(self._base_block_partitions), (
+            tasks,
+            self._base_block_partitions,
+        )
+        assert len(tasks) == len(self._fetched_metadata), (
+            tasks,
+            self._fetched_metadata,
+        )
+
+    def get_metadata(self, fetch_if_missing: bool = False) -> List[BlockMetadata]:
+        """Get the metadata for all blocks."""
+        if all(meta is not None for meta in self._fetched_metadata):
+            # Always return fetched metadata if we already have it.
+            metadata = self._fetched_metadata
+        elif not fetch_if_missing:
+            metadata = [
+                m if m is not None else t.get_metadata()
+                for m, t in zip(self._fetched_metadata, self._tasks)
+            ]
+        else:
+            _, metadata = self._get_blocks_with_metadata()
+        return metadata
+
+    def stats(self) -> DatasetStats:
+        """Create DatasetStats for this LazyBlockList."""
+        return DatasetStats(
+            stages={self._name: self.get_metadata(fetch_if_missing=False)},
+            parent=None,
+            needs_stats_actor=True,
+            stats_uuid=self._stats_uuid,
+        )
+
+    def _submit_task(self, task_idx: int) -> ObjectRef[MaybeBlockPartition]:
+        """Submit the task with index task_idx.
+
+        This will apply the pushdown function and use the base partition, if it exists.
+        """
+        if not self._execution_started:
+            stats_actor = _get_or_create_stats_actor()
+            stats_actor.record_start.remote(self._stats_uuid, self._name)
+            self._execution_started = True
+        task = self._tasks[task_idx]
+        if self._pushdown_fn is not None:
+            # If pushdown exists, compose pushdown function with read function.
+            base_part = self._base_block_partitions[task_idx]
+            pushdown_fn = self._pushdown_fn
+            if base_part is not None:
+                # If base partition exists, use it instead of reinvoking read function.
+                context = DatasetContext.get_current()
+
+                def read_fn():
+                    # TODO(Clark): Add base_part as an optional task dependency.
+                    part = ray.get(base_part)
+                    if context.block_splitting_enabled:
+                        # If dynamic block splitting enabled, fetch individual blocks
+                        # from within the partition and feed each block to the pushdown
+                        # function.
+                        for block_and_meta in part:
+                            block, _ = block_and_meta
+                            block = ray.get(block)
```

Comment:
An alternative here is to fetch the blocks in bulk:
```python
blocks, _ = zip(*block_and_meta)
for block in ray.get(list(blocks)):
    yield from pushdown_fn(part)
```
However, that would require materializing all of the blocks in the object store at once, while this streaming get would allow us to release the Plasma memory for a block after it has been processed.